### PR TITLE
Improving Linting Script with Error Handling and Configuration and Path Exceptions

### PR DESCRIPTION
**Pull request by AnandPolamarasetti**
There are a number of changes that have been made to the linting script in the recent past and these changes bring a lot of improvement to the script. 
 
 First of all, the script has been supplemented with error handling and reporting options of the new level. This enhancement makes it possible to address problems that may come up during the linting process in a better way by providing detailed information on the problems faced. New features of logging have been implemented to help users know the exact actions to perform to correct the code problems and in a more faster way. 
 
 As for the code analysis, the script has become more diverse and is capable of working with a number of file types and extensions. It uses `cpplint` for C++ files and `pylint` for Python files so as to cover all the possible linting needs. This improvement helps the script to overcome the deficiencies that would normally arise due to different code bases and follow the coding conventions of C++ and Python. 
 
 A new and marked enhancement is the availability of configuration options which can be set. New in version 1. 5. 0: Users are now able to specify a custom Pylint configuration file using the `--pylint-rc` option. This feature is an enhancement that increases the level of flexibility to the linting process by enabling users to configure the linting rules as they wish. 
 
 There are also improvements in the header guard convention in the update especially for DMLC projects. The script has been enhanced with a better way of generating header guards to enhance on the naming convention. This is very important while working with code especially in large projects, to ensure that the code is clean and there are no conflicts. 
 
 Also, the script has been updated with the ability of path exclusion. It is possible to skip certain paths and it is very helpful when you are working with some files that are not needed in a certain linting task. This enhances the effectiveness of linting as it eliminates or reduces the amount of code that needs to be checked to the important ones. 
 
 Last but not the least; there are improvements made on the script to make it runnable on virtually any operating system and Python version. Some changes have been made for file paths and encoding problems to provide better and smooth experience for users. 
 
 In a nutshell, these changes enhance the functionality of the linting script and make it a very useful tool in enforcing code standards across different development environments.

### CODE REVIEW

**Comment by AnandPolamarasetti** [comment]
This update significantly enhances the linting script with several key improvements:This update significantly enhances the linting script with several key improvements: 
 
 1. **Advanced Error Handling**: Some new features have been included in the script in order to avoid the occurrences of errors. This includes the improved encoding management as well as enhanced and meaningful error messages during the linting phase. 
 
 2. **Custom Configuration Support**: New option ––pylint-rc for the command line interface to supports the custom configuration files. New feature is the ability for users to choose their Pylint configuration file through which they can set the rules that they want to be checked during the linting process. 
 
 3. **Path Exclusions**: Added the feature for ignoring certain paths and all the folders within them while linting by using the `--exclude_path` option. This feature also assist in exclusion of files or directories that may not be of interest to the current linting process. 
 
 4. **Enhanced File Handling**: Improved the file processing algorithm to manage the files in a better way according to their types. The script has also been improved to work with more flexible file type as well as to normalize the file path. 
 
 5. **Improved Summary Reporting**: The summary reporting has been enhanced to better explain what kind of linting information it provides; what categories it covers and how many issues it has detected. 
 
 These improvements are meant to allow for easier linting, faster linting, and linting that is more suited to different projects.
